### ISSUE TITLE
Sharded Weight Files Not Uploaded by `_push_from_checkpoint` function in DeepSpeed ZeRO Stage 3 Training

### ISSUE THREAD

**Issue by e-mon** [huggingface/transformers#26140]
### System Info

- `transformers` version: 4.33.1
- Platform: Linux-4.18.0-372.9.1.el8.x86_64-x86_64-with-glibc2.28
- Python version: 3.10.10
- Huggingface_hub version: 0.17.1
- Safetensors version: 0.3.3
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes, deepspeed

### Who can help?

@muellerz and @pacman100

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I attempted to train "https://huggingface.co/codellama/CodeLlama-7b-hf" using DeepSpeed ZeRO Stage 3, but found that the `_push_from_checkpoint` function was not working properly, and the weights file was not being uploaded. I was able to confirm that the sharded weights file, named like `pytorch_model-00001-of-00002.bin`, was being saved in the checkpoint directory.

From what I can tell after looking at the code below, it does not seem that such divided files are included in the upload target. Is this the expected behavior? If the divided files should be included in the target, I am considering creating a Pull Request.
https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3624

### Expected behavior

all checkpoints to be uploaded to the Hugging Face Hub.

**Comment by pacman100** [comment]
Yes, you are correct. Currently, the sharded weights aren't being copied to the main output directory and as such not being uploaded to the Hub. It would be great if you could submit a PR resolving it.

**Comment by pacman100** [comment]
Hello @ArthurZucker, this is not limited to DeepSpeed, even when doing normal training with say DDP, if the model gets sharded, it won't be uploaded to the hub at present.
